### PR TITLE
Bug fix - extension array with 2d datetime64

### LINKED ISSUE #38390: BUG: ExtensionArray with 2D datetime64 values errors on display formatting
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import numpy as np
import pandas as pd
import pandas.io.formats.format as fmt
x = pd.date_range('2018-01-01', periods=5, freq='H').to_numpy()
x = np.tile(x, (3, 1))
formatter = fmt.Datetime64Formatter(x)
result = formatter.get_result()
```

#### Problem description

When using a Pandas ExtensionArray that has 2D values of Datetime64 dtype, Pandas errors on formatting the values with the following:
```
File "/ib/python3.8/site-packages/pandas/io/formats/printing.py", line 73, in <listcomp>
    return [x.rjust(max_len) for x in texts]
AttributeError: 'list' object has no attribute 'rjust'
```

The above is also true when using the `Datetime64Formatter` directly with values of ndim > 1.

#### Expected Output
Formatting that takes into account dimensionality:
```
[' [2018-01-01 00:00:00, 2018-01-01 01:00:00, 201...', ' [2018-01-01 05:00:00, 2018-01-01 06:00:00, 201...']
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : db08276bc116c438d3fdee492026f8223584c477
python           : 3.8.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-54-generic
Version          : #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.3
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.6.0.post20201009
Cython           : None
pytest           : 6.1.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 1.0.1
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by weikhor**
- [x] closes #38390
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -1622,17 +1622,30 @@ def __init__(
...
-        )
-        return fmt_values.tolist()
+        if len(values_shape) > 1:
+            fmt_values = np.reshape(values, values_shape)
+            nested_formatter = GenericArrayFormatter(fmt_values)
+            fmt_values = nested_formatter.get_result()
+        else:
+            fmt_values = values
+
+        if isinstance(fmt_values, np.ndarray):
```
Comment (jreback) [inline pandas/io/formats/format.py:1645]: what else can these be? can we just `list(fmt_values)`?

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -1810,11 +1823,23 @@ class Datetime64TZFormatter(Datetime64Formatter):
...
         values = self.values.astype(object)
+        values_shape = values.shape
+        values = values.ravel() if values.ndim > 1 else values
         ido = is_dates_only(values)
         formatter = self.formatter or get_format_datetime64(
             ido, date_format=self.date_format
         )
-        fmt_values = [formatter(x) for x in values]
+
+        if len(values_shape) > 1:
```
Comment (jreback) [inline pandas/io/formats/format.py:1833]: might be worth making a helper function ofr L1833-1842 so can call here & in the super method

**Review by jreback — CHANGES_REQUESTED** [review]
does this change user visible formatting or just with the internal formatters?

**Comment by weikhor** [comment]
> does this change user visible formatting or just with the internal formatters?
Hi code changes with the internal formatters

File: pandas/io/formats/format.py (hunk 1 above)
Comment (weikhor) [reply pandas/io/formats/format.py:1645]: Hi Yes. It can. Have updated the code.

File: pandas/io/formats/format.py [hunk 3]
```diff
@@ -1622,17 +1622,31 @@ def __init__(
...
+        values = values.ravel() if values.ndim > 1 else values
 
         if not isinstance(values, DatetimeIndex):
             values = DatetimeIndex(values)
 
         if self.formatter is not None and callable(self.formatter):
-            return [self.formatter(x) for x in values]
+            values = [self.formatter(x) for x in values]
+        else:
+            values = values._data._format_native_types(
```
Comment (jbrockmendel) [inline pandas/io/formats/format.py:1634]: The relevant methods should already support 2D, so i think we only need to fix/change the self.formatter line above?

Comment (weikhor) [reply pandas/io/formats/format.py:1634]: Ok Will change the self.formatter

**Comment by weikhor** [comment]
Hi, I am still stuck in the PR. Need more time.

File: pandas/io/formats/format.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/io/formats/format.py:1634]: might be worth looking at np.frompyfunc, ive been looking at using this as a replacement for ravel_compat to handle 2D cases

Comment (weikhor) [reply pandas/io/formats/format.py:1634]: Hi I am in loss to understand this.  After run steps inside Pandas code, step go into this function
In the pandas\core\arrays\datetimes.py,
```
@dtl.ravel_compat
def _format_native_types(
    self, *, na_rep="NaT", date_format=None, **kwargs
) -> npt.NDArray[np.object_]:
    from pandas.io.formats.format import get_format_datetime64_from_values

    fmt = get_format_datetime64_from_values(self, date_format)

    return tslib.format_array_from_datetime(
        self.asi8, tz=self.tz, format=fmt, na_rep=na_rep
    )
```
Are you mean to replace this function from ravel_compat  to np.frompyfunc in this file? Hope can give some guidances. Thank. On my side, I try best to figure this out.

File: pandas/tests/io/formats/test_format.py [hunk 4]
```diff
@@ -3158,6 +3158,90 @@ def format_func(x):
...
+            return t.strftime("%H-%m")
+
+        formatter = fmt.Datetime64Formatter(x.reshape((4, 2, 3)), formatter=format_func)
+        result = formatter.get_result()
+        assert len(result) == 4
+        assert result[0].strip() == "[[00-01, 01-01, 02-01], [03-01, 04-01, 05-01]]"
+        assert result[3].strip() == "[[18-01, 19-01, 20-01], [21-01, 22-01, 23-01]]"
+
+
+class TestDatetime64TZFormatter:
```
Comment (jbrockmendel) [inline pandas/tests/io/formats/test_format.py:3198]: are any of these user-facing?  e.g. is there some DataFrame method that could trigger the bug that this fixes?

File: pandas/io/formats/format.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/io/formats/format.py:1634]: my thought was that L1632 might end up looking something like `values = np.frompyfunc(self.formatter, 1, 1)(values)`

File: pandas/tests/io/formats/test_format.py (hunk 4 above)
Comment (weikhor) [reply pandas/tests/io/formats/test_format.py:3198]: Hi, no able to trigger other bug from this fix.

File: pandas/io/formats/format.py (hunk 3 above)
Comment (weikhor) [reply pandas/io/formats/format.py]: Hi, I have try to change code to 
```
def _format_strings(self) -> list[str]:
        """we by definition have DO NOT have a TZ"""

        values = self.values

        if not isinstance(values, DatetimeIndex):
            values = DatetimeIndex(values)

        if self.formatter is not None and callable(self.formatter):
            fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
        else:
            fmt_values = values._data._format_native_types(
                na_rep=self.nat_rep, date_format=self.date_format
            )

        nested_formatter = GenericArrayFormatter(fmt_values)
        fmt_values = nested_formatter.get_result()
        return list(fmt_values)
```
I test this.
```
x = date_range("2018-01-01", periods=24, freq="H").to_numpy()

def format_func(t):
    return t.strftime("%H-%m")

formatter = fmt.Datetime64Formatter(x.reshape((4, 2, 3)), formatter=format_func)
result = formatter.get_result()
```

The error looks like this
```
  from pandas import (
Traceback (most recent call last):
  File "D:\kcw\development\testing.py", line 73, in <module>
    result = formatter.get_result()
  File "d:\kcw\pandas\pandas\io\formats\format.py", line 1349, in get_result
    fmt_values = self._format_strings()
  File "d:\kcw\pandas\pandas\io\formats\format.py", line 1631, in _format_strings
    fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
  File "d:\kcw\pandas\pandas\core\indexes\base.py", line 949, in __array_ufunc__
    result = getattr(ufunc, method)(*new_inputs, **kwargs)
  File "d:\kcw\pandas\pandas\core\arrays\datetimelike.py", line 1793, in __array_ufunc__
    return super().__array_ufunc__(ufunc, method, *inputs, **kwargs)
  File "d:\kcw\pandas\pandas\core\arrays\base.py", line 1648, in __array_ufunc__
    return arraylike.default_array_ufunc(self, ufunc, method, *inputs, **kwargs)  File "d:\kcw\pandas\pandas\core\arraylike.py", line 490, in default_array_ufunc
    return getattr(ufunc, method)(*new_inputs, **kwargs)
  File "D:\kcw\development\testing.py", line 70, in format_func
    return t.strftime("%H-%m")
AttributeError: 'int' object has no attribute 'strftime'
```
But I encounter this problem. In this line `fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)` . The np.frompyfunc function converts the values <class 'pandas.core.indexes.datetimes.DatetimeIndex'> to integer (ns).

Comment (weikhor) [reply pandas/io/formats/format.py]: Hi. 
1. For this class `Datetime64Formatter`,  
```
x = date_range("2018-01-01", periods=10, freq="H").to_numpy()
values = DatetimeIndex(x)
fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
def format_func(t):
return t.strftime("%H-%m")

The np.frompyfunc will convert `2018-01-01` to ns integer value which

**Comment by jreback** [comment]
@jbrockmendel ok here?

@weikhor if you can merge master

**Comment by jbrockmendel** [comment]
will take another look today or tomorrow

File: pandas/io/formats/format.py [hunk 5]
```diff
@@ -1621,18 +1621,23 @@ def __init__(
...
 
-        fmt_values = values._data._format_native_types(
-            na_rep=self.nat_rep, date_format=self.date_format
-        )
-        return fmt_values.tolist()
+        if fmt_values.ndim > 1:
+            nested_formatter = GenericArrayFormatter(fmt_values)
+            return list(nested_formatter.get_result())
+
+        return list(fmt_values)
```
Comment (jbrockmendel) [inline pandas/io/formats/format.py:1640]: why is _anything_ necessary between L1630-L1640?  we know we have DatetimeIndex which is 1D.

File: pandas/io/formats/format.py [hunk 6]
```diff
@@ -1621,18 +1621,23 @@ def __init__(
 
     def _format_strings(self) -> list[str]:
         """we by definition have DO NOT have a TZ"""
-        values = self.values
 
+        values = self.values
         if not isinstance(values, DatetimeIndex):
             values = DatetimeIndex(values)
```
Comment (jbrockmendel) [inline pandas/io/formats/format.py:1627]: OK, so there is a deeper bug here in that DatetimeIndex(values) isn't raising when passed a 2D array.

File: pandas/io/formats/format.py [hunk 7]
```diff
@@ -1621,18 +1621,23 @@ def __init__(
 
     def _format_strings(self) -> list[str]:
         """we by definition have DO NOT have a TZ"""
-        values = self.values
```
Comment (jbrockmendel) [inline pandas/io/formats/format.py:1624]: what is the scenario in which we get here with non-1D values?  could we just assert 1d-ness at the top and not worry about the rest?

**Comment by weikhor** [comment]
Hi It is difficult for me to answer questions provided. I need more time.

**Comment by weikhor** [comment]
I close this since I stuck this for long time. I get simpler task. Thank for review.

**Pull request closed without merging**